qiskit_machine_learning/kernels/statevector_kernel.py

```diff
@@ -0,0 +1,190 @@
+
+    Here, the kernel function is defined as the overlap of two quantum states defined by a
+    parametrized quantum circuit (called feature map):
+
+    .. math::
+
+        K(x,y) = |\langle \phi(x) | \phi(y) \rangle|^2
+
+    In this implementation, :math:`\phi` is represented by a ``Statevector.data`` array,
+    thus the kernel function is simply:
+
+    .. math::
+
+        K(x,y) = (\phi(x)^\dagger \phi(y))^2
+
+    These are stored in a statevector cache for reuse to avoid repeated computation. This stash
+    can be cleared using :meth:`clear_cache()`.
+    """
+
+    def __init__(
+        self,
+        *,
+        feature_map: QuantumCircuit | None = None,
+        enforce_psd: bool = True,
+        evaluate_duplicates: str = "off_diagonal",
+    ) -> None:
+        """
+        Args:
+            feature_map: Parameterized circuit to be used as the feature map. If ``None`` is given,
+                :class:`~qiskit.circuit.library.ZZFeatureMap` is used with two qubits. If there's
+                a mismatch in the number of qubits of the feature map and the number of features
+                in the dataset, then the kernel will try to adjust the feature map to reflect the
+                number of features.
+            enforce_psd: Project to the closest positive semidefinite matrix if ``x = y``.
+                Default ``True``.
+            evaluate_duplicates: Defines a strategy how kernel matrix elements are evaluated if
+               duplicate samples are found. Possible values are:
+
+                    - ``all`` means that all kernel matrix elements are evaluated, even the diagonal
+                      ones when training. This may introduce additional noise in the matrix.
+                    - ``off_diagonal`` when training the matrix diagonal is set to `1`, the remaining
+                      elements are fully evaluated, e.g., for two identical samples in the
+                      dataset. When inferring, all elements are evaluated. This is the default
+                      value.
+                    - ``none`` when training the diagonal is set to `1` and if two identical samples
+                      are found in the dataset the corresponding matrix element is set to `1`.
+                      When inferring, matrix elements for identical samples are set to `1`.
+
+        Raises:
+            ValueError: When unsupported value is passed to `evaluate_duplicates`.
+        """
+        super().__init__(feature_map=feature_map, enforce_psd=enforce_psd)
+
+        eval_duplicates = evaluate_duplicates.lower()
+        if eval_duplicates not in ("all", "off_diagonal", "none"):
+            raise ValueError(
+                f"Unsupported value passed as evaluate_duplicates: {evaluate_duplicates}"
+            )
+        self._evaluate_duplicates = eval_duplicates
+
+        self._statevector_cache = {}
+
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray | None = None) -> np.ndarray:
```

Comment:
As a different thought on the cache you could look at the functools caching decorators e.g  lru_cache which I think could be applied to something like your get statevector method so that you do not have to do explicit caching - an lru cache can be size limited so would alleviate concerns about uncontrolled growth and the user needing to manage things via clear cache.

Reply:
> I'd hide cache in the implementation without exposing anything and clear it once `evaluate` is called.

This should definitely be optional. For example in `PegasosQSVC` evaluate is called in every step of the training loop, so there it would be crucial to keep the cache.

Reply:
> > I'd hide cache in the implementation without exposing anything and clear it once `evaluate` is called.
> 
> This should definitely be optional. For example in `PegasosQSVC` evaluate is called in every step of the training loop, so there it would be crucial to keep the cache.

Not necessarily. `Pegasos` randomly samples elements to compute, so cache should be on the algorithm level, not on the kernel level. Moreover, current implementation does not have any cache anyway :)

Reply:
> Not necessarily. `Pegasos` randomly samples elements to compute, so cache should be on the algorithm level, not on the kernel level. Moreover, current implementation does not have any cache anyway :)

`Pegasos` samples a new point every iteration and then computes the kernel entries between this point and the (iteratively accumulated) support vectors. So the in a state vector based implementation of the kernel, the state vectors of the support vectors are used in every iterations and caching them would provide a speedup.

Of course so far there is no caching, but if caching is added I think limiting it to a single `evaluate` call might be too restrictive.

Reply:
Yes, `Pegasos` knows what should be cached, but `Pegasos` only has access to the kernel entries, not to the actual state vectors. So, unless you suggest writing a separate `StatevectorPegasos` that doesn't rely on the kernel interface, the caching has to be implemented within the kernel (or even better in the fidelity instance, but I accept that this is too challenging).

Reply:
Okay, thanks for being patient with me. I see the problem now. Certainly, I'm against `StatevectorPegasos`, but on the other hand I don't like the idea of exposing a caching interface at the `FidelityStatevectorKernel` level as well. Looks like caching makes sense only in the statevector kernel implementation. Any ideas where else any kind of caching may emerge? I'm thinking what bits and pieces can be unified.